go/plugins/internal/gemini/gemini.go

```diff
@@ -26,3 +40,475 @@ var (
+		for _, t := range tools {
+			toolNames = append(toolNames, t.Name)
+		}
+	}
+	return &genai.ToolConfig{
+		FunctionCallingConfig: &genai.FunctionCallingConfig{
+			Mode:                 mode,
+			AllowedFunctionNames: toolNames,
+		},
+	}
+}
+
+// translateCandidate translates from a genai.GenerateContentResponse to an ai.ModelResponse.
+func translateCandidate(cand *genai.Candidate) *ai.ModelResponse {
+	m := &ai.ModelResponse{}
+	switch cand.FinishReason {
+	case genai.FinishReasonStop:
+		m.FinishReason = ai.FinishReasonStop
+	case genai.FinishReasonMaxTokens:
+		m.FinishReason = ai.FinishReasonLength
+	case genai.FinishReasonSafety:
+		m.FinishReason = ai.FinishReasonBlocked
+	case genai.FinishReasonRecitation:
+		m.FinishReason = ai.FinishReasonBlocked
+	case genai.FinishReasonOther:
+		m.FinishReason = ai.FinishReasonOther
+	default: // Unspecified
+		m.FinishReason = ai.FinishReasonUnknown
+	}
+	msg := &ai.Message{}
+	msg.Role = ai.Role(cand.Content.Role)
+
+	// iterate over the candidate parts, only one struct member
+	// must be populated, more than one is considered an error
+	for _, part := range cand.Content.Parts {
+		var p *ai.Part
+		partFound := 0
+
+		if part.Text != "" {
+			partFound++
+			p = ai.NewTextPart(part.Text)
+		}
+		if part.InlineData != nil {
+			partFound++
+			p = ai.NewMediaPart(part.InlineData.MIMEType, string(part.InlineData.Data))
+		}
+		if part.FunctionCall != nil {
+			partFound++
+			p = ai.NewToolRequestPart(&ai.ToolRequest{
+				Name:  part.FunctionCall.Name,
+				Input: part.FunctionCall.Args,
+			})
+		}
+		if partFound > 1 {
+			panic(fmt.Sprintf("expected only 1 content part in response, got %d, part: %#v", partFound, part))
+		}
+
+		msg.Content = append(msg.Content, p)
+	}
+	m.Message = msg
+	return m
+}
+
+// Translate from a genai.GenerateContentResponse to a ai.ModelResponse.
+func translateResponse(resp *genai.GenerateContentResponse) *ai.ModelResponse {
+	r := translateCandidate(resp.Candidates[0])
+
+	r.Usage = &ai.GenerationUsage{}
+	if u := resp.UsageMetadata; u != nil {
+		r.Usage.InputTokens = int(*u.PromptTokenCount)
+		r.Usage.OutputTokens = int(*u.CandidatesTokenCount)
+		r.Usage.TotalTokens = int(u.TotalTokenCount)
+	}
+	return r
+}
+
+// convertParts converts a slice of *ai.Part to a slice of genai.Part.
+func convertParts(parts []*ai.Part) ([]*genai.Part, error) {
+	res := make([]*genai.Part, 0, len(parts))
+	for _, p := range parts {
+		part, err := convertPart(p)
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, part)
+	}
+	return res, nil
+}
+
+// convertPart converts a *ai.Part to a genai.Part.
+func convertPart(p *ai.Part) (*genai.Part, error) {
+	switch {
+	case p.IsText():
+		return genai.NewPartFromText(p.Text), nil
+	case p.IsMedia():
+		contentType, data, err := uri.Data(p)
+		if err != nil {
+			return nil, err
+		}
+		return genai.NewPartFromBytes(data, contentType), nil
+	case p.IsData():
+		panic("data parts not supported")
```

Comment:
Is this still true?

Reply:
Apparently yes (https://github.com/googleapis/go-genai/blob/8a615165d2161f5be0efb0d7bf5f77570166b0b0/types.go#L432)

Reply:
Isn't it this? https://github.com/googleapis/go-genai/blob/8a615165d2161f5be0efb0d7bf5f77570166b0b0/types.go#L451